File: mobile-client/android/Data/System/Security/RequestSignatureApplier.kt
package com.smartresponsor.mobile.client.data.system.security

import com.smartresponsor.mobile.client.contract.system.security.signatureheaders
import okhttp3.Request
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

class RequestSignatureApplier(private val secret: String) {
    fun sign(data: String): String {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(secret.toByteArray(), "HmacSHA256"))
        return mac.doFinal(data.toByteArray()).joinToString("") { "%02x".format(it) }
    }

    fun headersFor(path: String): SignatureHeaders = SignatureHeaders(signatureValue = sign(path))

    fun apply(request: Request): Request {
        val headers = headersFor(request.url.encodedPath)
        return request.newBuilder().header(headers.signatureHeaderName, headers.signatureValue).build()
    }
}

File: client/ios/Data/System/Security/RequestSignatureApplier.swift
import CryptoKit
import Foundation

public struct RequestSignatureApplier {
    private let secret: String

    public init(secret: String) {
        self.secret = secret
    }

    public func headers(for path: String) -> SignatureHeaders {
        SignatureHeaders(signatureValue: sign(data: path))
    }

    public func apply(to request: URLRequest) -> URLRequest {
        var copy = request
        let path = request.url?.path ?? "/"
        let headers = headers(for: path)
        copy.setValue(headers.signatureValue, forHTTPHeaderField: headers.signatureHeaderName)
        return copy
    }

    private func sign(data: String) -> String {
        let key = SymmetricKey(data: Data(secret.utf8))
        let signature = HMAC<SHA256>.authenticationCode(for: Data(data.utf8), using: key)
        return Data(signature).map { String(format: "%02x", $0) }.joined()
    }
}
